### PR TITLE
perf: memoize db_engine_spec in database

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
A recent PR #14547 introduced a performance regression causing dataset metadata fetching to become very slow for datasets with large numbers of columns. I originally thought the type regexes were the problem, but when researching the problem more closely it turns out that just referencing `self.table.database.db_engine_spec` in a `TableColumn` instance cost ~6ms on my local machine. Multiply that by 1000 columns ~= 6000 ms. To get around this I added memoization to the semi-expensive regex, but also added memoizing for `Database.db_engine_spec`. This should also speed up query rendering a bit, as there was similar logic there.

### BEFORE #14547 (pre-regression)
For the World Bank dataset (328 cols), fetching the data took slightly less than 180ms before on my local machine (including the unnecessary 20 ms redirect):
image

### CURRENT (master)
For the same dataset, retrieval of data now takes ~10s!
image

### AFTER
Retrieval is now slightly quicker than originally (including no redirect):
image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by kgabryje** [comment]
Can we remove the cypress timeout overrides introduced in the other PR?

**Comment by villebro** [comment]
> Can we remove the cypress timeout overrides introduced in the other PR?

Good idea; I'll do that. I've also been looking into introducing perf tests on the backend to identify these easier (will be a follow-up PR).

File: superset/connectors/sqla/models.py [hunk 1]
```diff
@@ -231,14 +222,20 @@ def is_temporal(self) -> bool:
...
-            return False
-        return column_spec.is_dttm
+        return self.type_generic == GenericDataType.TEMPORAL
+
+    @property
+    def db_engine_spec(self) -> Type[BaseEngineSpec]:
+        return self.table.db_engine_spec
+
+    @property
+    def type_generic(self) -> Optional[utils.GenericDataType]:
```
Comment (dpgaspar) [inline superset/connectors/sqla/models.py:232]: consider renaming to `get_generic_type` gives an head warning that it will do some computation around it

**Review by dpgaspar — APPROVED** [review]

File: superset/connectors/sqla/models.py (hunk 1 above)
Comment (villebro) [reply superset/connectors/sqla/models.py:232]: This is added to the column payload in the dataset request to complement the existing `type` field, so I think we need to keep it as a property.

**Pull request merged**